### PR TITLE
Fix logic error in FF_LockDirectory and FF_LockFAT

### PR DESCRIPTION
**Pull request by carlk3**
Reverse erroneous changes introduced in
https://github.com/FreeRTOS/Lab-Project-FreeRTOS-FAT/compare/c50b4b9..97e608c
at my instigation.

I had a misunderstanding of how `xEventGroupWaitBits` works, and tried to oversimplify the code. The lock must be done in two atomic steps, as in the original. 

In `FF_LockFAT`, for example, `xEventGroupWaitBits` and `xEventGroupClearBits` must be done as two separate steps. There is a meaningful difference between

```
		xEventGroupWaitBits( pxIOManager->xEventGroup,
			FF_FAT_LOCK_EVENT_BITS, /* uxBitsToWaitFor */
			pdFALSE,                /* xClearOnExit */
			pdFALSE,                /* xWaitForAllBits n.a. */
			pdMS_TO_TICKS( 10000UL ) );
		xBits = xEventGroupClearBits( pxIOManager->xEventGroup, FF_FAT_LOCK_EVENT_BITS );
```

and

```
        xBits = xEventGroupWaitBits( pxIOManager->xEventGroup,
                                     FF_FAT_LOCK_EVENT_BITS, /* uxBitsToWaitFor */
                                     pdTRUE,                 /* xClearOnExit */
                                     pdFALSE,                /* xWaitForAllBits n.a. */
                                     FF_TIME_TO_WAIT_FOR_EVENT_TICKS );
```

When `xEventGroupSetBits( pxIOManager->xEventGroup, FF_FAT_LOCK_EVENT_BITS )` is called, it will unblock all tasks waiting on `FF_FAT_LOCK_EVENT_BITS` in `xEventGroupWaitBits`. If `xClearOnExit` is `pdTRUE`, they will all clear the bit, but that doesn’t help anything. `xEventGroupClearBits` needs to be done as a second atomic step. Only a single task will get back `FF_FAT_LOCK_EVENT_BITS` in `xBits`, and after that task has atomically cleared the bit, all the other tasks will get back zero. So, only one task succeeds in the following test:

```
        if( ( xBits & FF_FAT_LOCK_EVENT_BITS ) != 0 )
        {
            /* This task has cleared the desired bit.
             * It now 'owns' the resource. */
            pxIOManager->pvFATLockHandle = xTaskGetCurrentTaskHandle();
            break;
        }
```

and the others just loop back around to wait again.

I have a project ([carlk3/FreeRTOS-FAT-CLI-for-RPi-Pico ](https://github.com/carlk3/FreeRTOS-FAT-CLI-for-RPi-Pico)) running on the official SMP FreeRTOS RP2040 port ([FreeRTOS/FreeRTOS-Kernel](https://github.com/FreeRTOS/FreeRTOS-Kernel/tree/smp/portable/ThirdParty/GCC/RP2040)) on [Raspberry Pi Pico](https://www.raspberrypi.org/documentation/microcontrollers/raspberry-pi-pico.html), and this error causes failures such as

`assertion "( xEventGroupGetBits( pxIOManager->xEventGroup ) & FF_FAT_LOCK_EVENT_BITS ) == 0" failed: file "/home/carlk/pi/pico/FreeRTOS+FAT+CLI/Lab-Project-FreeRTOS-FAT/ff_locking.c", line 297, function: FF_UnlockFAT`

when running standard tests such as the old `vMultiTaskStdioWithCWDTest` from `FreeRTOS-Plus/FreeRTOS-Plus/Demo/Common/FreeRTOS_Plus_FAT_Demos/test/ff_stdio_tests_with_cwd.c`.

### CODE REVIEW

File: ff_locking.c [hunk 1]
```diff
@@ -163,18 +167,19 @@ void FF_LockDirectory( FF_IOManager_t * pxIOManager )
...
-         * It waits for the desired bit to come high, and clears the
-         * bit so that other tasks can not take it. */
-        xBits = xEventGroupWaitBits( pxIOManager->xEventGroup,
-                                     FF_DIR_LOCK_EVENT_BITS, /* uxBitsToWaitFor */
-                                     pdTRUE,                 /* xClearOnExit */
-                                     pdFALSE,                /* xWaitForAllBits n.a. */
-                                     FF_TIME_TO_WAIT_FOR_EVENT_TICKS );
-
-        if( ( xBits & FF_DIR_LOCK_EVENT_BITS ) != 0 )
+        xEventGroupWaitBits(pxIOManager->xEventGroup,
```
Comment (htibosch) [inline ff_locking.c:172]: This extra statement was indeed necessary.

**Review by htibosch — APPROVED** [review]
Thanks a lot Carl for this PR.
I have written [a simulation](https://forums.freertos.org/t/xeventgroupwaitbits-broken-in-freertos-smp/14075/13) just to be sure that the problem also occurs on a single core platform, and it did.

So I approve this PR.

Thanks

**Review by AniruddhaKanhere — APPROVED** [review]

**Review by paulbartell — APPROVED** [review]

File: ff_locking.c (hunk 1 above)
Comment (paulbartell) [reply ff_locking.c:172]: The return value of xEventGroupWaitBits is not checked here. Is that on purpose?

if the call times out (FF_TIME_TO_WAIT_FOR_EVENT_TICKS ticks elapsed), then the event bit will be erroneously cleared right?

File: ff_locking.c [hunk 2]
```diff
@@ -264,21 +272,23 @@ void FF_LockFAT( FF_IOManager_t * pxIOManager )
...
-         * It waits for the desired bit to come high, and clears the
-         * bit so that other tasks can not take it. */
-        xBits = xEventGroupWaitBits( pxIOManager->xEventGroup,
-                                     FF_FAT_LOCK_EVENT_BITS, /* uxBitsToWaitFor */
-                                     pdTRUE,                 /* xClearOnExit */
-                                     pdFALSE,                /* xWaitForAllBits n.a. */
-                                     FF_TIME_TO_WAIT_FOR_EVENT_TICKS );
-
-        if( ( xBits & FF_FAT_LOCK_EVENT_BITS ) != 0 )
+        xEventGroupWaitBits(pxIOManager->xEventGroup,
```
Comment (paulbartell) [inline ff_locking.c:277]: Similar issuer here as well. The return value of xEventGroupWaitBits is not checked.

**Pull request merged**